golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -0,0 +1,384 @@
+
+// SnapshotFormat 1 is a proto message containing an artifact name, and the binary artifact data
+const SnapshotFormat = 1
+
+const ExportManifestFilename = "export-manifest.json"
+const ExportDataFilename = "export-data.json"
+const UntrustedExportDataArtifactName = "@@@### Untrusted Export Data ###@@@"
+const UntrustedExportDataFilename = "untrusted-export-data.json"
+const ExportedFilesMode = 0644
+
+type activeSnapshot struct {
+	height        int64
+	startedResult chan error
+	retrieved     bool
+}
+
+type exportManifest struct {
+	BlockHeight uint64      `json:"blockHeight,omitempty"`
+	Data        string      `json:"data,omitempty"`
+	Artifacts   [][2]string `json:"artifacts"`
+}
+
+type SwingStoreExporter interface {
+	ExportSwingStore(ctx sdk.Context) []*vstoragetypes.DataEntry
+}
+
+type SwingsetSnapshotter struct {
+	app            *baseapp.BaseApp
+	exporter       SwingStoreExporter
+	blockingSend   func(action vm.Jsonable) (string, error)
+	activeSnapshot *activeSnapshot
+}
+
+type snapshotAction struct {
+	Type        string            `json:"type"` // COSMOS_SNAPSHOT
+	BlockHeight int64             `json:"blockHeight"`
+	Request     string            `json:"request"` // "initiate", "discard", "retrieve", or "restore"
+	Args        []json.RawMessage `json:"args,omitempty"`
+}
+
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, exporter SwingStoreExporter, sendToController func(bool, string) (string, error)) SwingsetSnapshotter {
+	blockingSend := func(action vm.Jsonable) (string, error) {
+		bz, err := json.Marshal(action)
+		if err != nil {
+			return "", err
+		}
+		return sendToController(true, string(bz))
+	}
+
+	return SwingsetSnapshotter{
+		app:            app,
+		exporter:       exporter,
+		blockingSend:   blockingSend,
+		activeSnapshot: nil,
+	}
+}
+
+func (snapshotter *SwingsetSnapshotter) logger() log.Logger {
+	return snapshotter.app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "snapshotter")
+}
+
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if snapshotter.activeSnapshot != nil {
+		return fmt.Errorf("snapshot already in progress for height %d", snapshotter.activeSnapshot.height)
+	}
+
+	if snapshotter.app.SnapshotManager() == nil {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	snapshotter.activeSnapshot = &activeSnapshot{
+		height,
+		make(chan error, 1),
+		false,
+	}
+
+	go func(startedResult chan error) {
+		logger := snapshotter.logger()
+
+		action := &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotter.activeSnapshot.height,
+			Request:     "initiate",
+		}
+
+		_, err := snapshotter.blockingSend(action)
+
+		if err != nil {
+			snapshotter.activeSnapshot = nil
+			startedResult <- err
+			logger.Error("failed to initiate swingset snapshot", "err", err)
+			return
+		}
+
+		startedResult <- nil
+
+		snapshotter.app.Snapshot(height)
+
+		if !snapshotter.activeSnapshot.retrieved {
```

Comment:
Avoid deep nesting by returning as soon as possible:
```suggestion
		if snapshotter.activeSnapshot.retrieved {
		    snapshotter.activeSnapshot = nil
		    return
		}
```